subprojects/build-cache/src/main/java/org/gradle/caching/local/internal/DirectoryBuildCacheServiceFactory.java

```diff
@@ -96,6 +100,15 @@ public BuildCacheService createBuildCacheService(DirectoryBuildCache configurati
         return new DirectoryBuildCacheService(fileStore, persistentCache, tempFileStore, fileAccessTracker, FAILED_READ_SUFFIX);
     }
 
+    private CleanupAction createCleanupAction(final int removeUnusedEntriesAfterDays) {
+        return CacheCleanupFileAccessTimeProvidingCleanupAction.create(fileAccessTimeJournal, new Transformer<CleanupAction, CacheCleanupFileAccessTimeProvider>() {
+            @Override
+            public CleanupAction transform(CacheCleanupFileAccessTimeProvider fileAccessTimeProvider) {
```

Comment:
I find this rather hard to understand, so let me see if I got this right: Instead of injecting the journal into the cleanup action, we instead decorate it, pass it back and then construct the cleanup action using that decorator. 

Would this be simplified if the journal was an argument to `clean` instead? Then there would not be a double decoration, but only one layer.

Reply:
I tried changing the `CleanupAction` interface at first but that didn't work out because then `CacheAccess` would have to know about `FileAccessTimeJournal` but the latter is user-home-scoped. I guess we could add a subclass of CleanupAction that takes a `CacheCleanupFileAccessTimeProvider` and change `CacheCleanupFileAccessTimeProvidingCleanupAction` to take a delegate of it.